modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -500,6 +545,189 @@ public int getZoneId(String name) {
+                        }
+
+                        if (topVerScaleUpFut0 == null) {
+                            topVerScaleUpFut0 = new CompletableFuture<>();
+
+                            zoneState.revisionScaleUpFutures().put(scaleUpRevision, topVerScaleUpFut0);
+                        }
+
+                        topVerScaleUpFut0.handle((ignored0, e) -> {
+                            if (e == null) {
+                                topVerScaleUpFut.complete(null);
+                            } else {
+                                topVerScaleUpFut.completeExceptionally(e);
+                            }
+
+                            return null;
+                        });
+                    } else {
+                        topVerScaleUpFut.complete(null);
+                    }
+                } else {
+                    topVerScaleUpFut.complete(null);
+                }
+            }
+        });
+
+        CompletableFuture<Void> topVerScaleDownFut = new CompletableFuture<>();
+
+        topVerFut.thenAccept(ignored -> {
+            CompletableFuture<Void> topVerScaleDownFut0 = null;
+
+            synchronized (dataNodesMutex) {
+                ZoneState zoneState = zonesState.get(zoneId);
+
+                if (zoneState == null) {
+                    throw new DistributionZoneNotFoundException(zoneId);
+                }
+
+                if (immediateScaleDown.get()) {
+                    Map.Entry<Long, Long> scaleDownRevisionEntry = topVerAndScaleDownRevision.ceilingEntry(topVer);
+
+                    Long scaleDownRevision = null;
+
+                    if (scaleDownRevisionEntry != null) {
+                        scaleDownRevision = scaleDownRevisionEntry.getValue();
+                    }
+
+                    if (scaleDownRevision != null && zoneState.scaleDownRevision() < scaleDownRevision) {
+                        Map.Entry<Long, CompletableFuture<Void>> ceilingEntry =
+                                zoneState.revisionScaleDownFutures().ceilingEntry(scaleDownRevision);
+
+                        if (ceilingEntry != null) {
+                            topVerScaleDownFut0 = ceilingEntry.getValue();
+                        }
+
+                        if (topVerScaleDownFut0 == null) {
+                            topVerScaleDownFut0 = new CompletableFuture<>();
+
+                            zoneState.revisionScaleDownFutures().put(scaleDownRevision, topVerScaleDownFut0);
+                        }
+
+                        topVerScaleDownFut0.handle((ignored0, e) -> {
+                            if (e == null) {
+                                topVerScaleDownFut.complete(null);
+                            } else {
+                                topVerScaleDownFut.completeExceptionally(e);
+                            }
+
+                            return null;
+                        });
+                    } else {
+                        topVerScaleDownFut.complete(null);
```

Comment:
What does this clause means? Please add the javadoc explaining why it's possible.

Reply:
I updated java doc. But I don't think this is the last java doc fix.